src/librustc_mir/build/mod.rs

```diff
@@ -341,6 +341,11 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             }
         }
     }
+
+    pub fn check_overflow(&self) -> bool {
```

Comment:
It seems to me that the library with the generic/inline methods should be defining whether it has overflow checks or not, independent of how it is used downstream.

Functions are supposed to be some sort of abstraction boundary, and it seems unfortunate that adding `#[inline]` would break that boundary (i.e. adding/removing it changes behaviour). Also, if we ever have scoped/per-crate overflow check control, all functions in a given scope/crate should be controlled by that, not whether they happen to be generic or not (etc.). As has been noted above, this is made more difficult by some functions wanting to be part of the overflow check abstraction itself (the proxy trait impls/`pow`/other arithmetic-like operations), but I don't think assuming that this is the case for all inline/generic functions is the right default.

Distributing debug libraries---at least, libraries with debug assertions and debuginfo enabled---doesn't seem like the worst thing, e.g. Haskell does it for profiling, and we may want it for this sort of thing too. Also, getting more aggressive/widespread coverage of `std`/`core`'s debug assertions seems like it would be good for everyone, although (ab)using all Rust users non-`--release` builds for this purpose isn't necessarily the best way to do it.